[WIP] draft - chunk merge files during cache write to avoid reading excessi…

sujay-jain opened the pull request:
chunk merge files during cache write to avoid reading excessively large data into memory at once

```
== RELEASE NOTES ==

```
== NO RELEASE NOTE ==
```

kewang1024 committed:
Enable creating table with table_supports_delta_delete property

Add table_supports_delta_delete in raptor metadata including Metadata
class, table property and system table. Create index table with
table_supports_delta_delete field. Add table_supports_delta_delete
in 'tables' table.

NikhilCollooru committed:
Support supplying benchmark queries through MySQL

Create a new module presto-benchmark-runner to
enable running benchmark queries on Presto.

bhhari committed:
Add support for filter functions on columns with coercions

Filter expressions are referring to columns from the latest table schema, hence,
variables in the variableToInput mapping in OrcSelectivePageSourceFactory must
use column types from the latest table schema, not from the partition schema.

shixuan-fan committed:
Use distinct table names in TestHiveRecoverableGroupedExecution

viczhang861 committed:
Throw PrestoException for unsupported correlated subquery

sujay-jain committed:
chunk merge files during cache write to avoid reading excessively large data into memory and other optimizations

highker commented on presto-cache/src/main/java/com/facebook/presto/cache/LocalRangeCacheManager.java:
```diff
@@ -261,17 +263,32 @@ private boolean write(FileReadRequest key, byte[] data, Path newFilePath)
             else {
                 // copy previous file's data to the new file
                 byte[] previousFileBytes = Files.readAllBytes(new File(previousCacheFile.getPath().toUri()).toPath());
```
rather than creating a large array `previousFileBytes` by reading everything from `previousCacheFile`, read the data to `previousCacheFile`  in small chunks and gradually append to `newFilePath`.

highker commented on presto-cache/src/main/java/com/facebook/presto/cache/LocalRangeCacheManager.java:
```diff
@@ -261,17 +263,32 @@ private boolean write(FileReadRequest key, byte[] data, Path newFilePath)
 
                 // remove the overlapping part and append the remaining cache data
-                byte[] remainingCacheFileBytes = new byte[toIntExact((key.getLength() + key.getOffset()) - (previousFileLength + previousFileOffset))];
-                System.arraycopy(data, toIntExact(previousFileLength + previousFileOffset - key.getOffset()), remainingCacheFileBytes, 0, remainingCacheFileBytes.length);
```
`data` is usually small; so we can append it to `newFilePath` once. But can we avoid an extra copy from `data` to `remainingCacheFileBytes` and then to `newFilePath`? Feel free to make it more efficient thought the original code isn't

sujay-jain replied:
this context regarding ```data``` size is useful :) Changed to using a ```RandomAccessFile``` instead which should avoid the extra array copy. What do you think?

highker commented on presto-cache/src/main/java/com/facebook/presto/cache/LocalRangeCacheManager.java:
```diff
@@ -261,17 +263,32 @@ private boolean write(FileReadRequest key, byte[] data, Path newFilePath)
+                newFileLength = previousFileLength + numBytesToCopy;
                 newFileOffset = previousFileOffset;
             }
 
```
Also need to use the "small copy" logic to the `if (followingCacheFile != null)` branch

viczhang861 committed:
Minor refactor of ApplyNode constructor

rongrong committed:
Fix SQL function compilation failure

RowExpression compilation for SQL function could fail when the input
is not referenced in function body.
